File: practice/practical-work7/Main.swift
//
// Object-oriented programing (college 4th course)
// Practice: Practical work 7
//

extension LosslessStringConvertible {
  /// Returns a value read from standard input.
  static func read(lable: String = "", nestingLevel: Int = 0) -> Self {
    let indent = String(repeating: " ", count: nestingLevel * 2)
    let decor = lable.isEmpty ? "\(indent)> " : "\(indent)> \(lable) = "

    while true {
      print(decor, terminator: "")
      if let rawValue = readLine() {
        if let value = Self(rawValue) {
          return value
        } else {
          print("Error: Value must be '\(type(of: self))'")
        }
      } else {
        fatalError("Current line or until EOF is reached")
      }
    }
  }
}

extension LosslessStringConvertible where Self: Comparable {
  /// Returns a value within the specified range read from standard input.
  static func read(in range: ClosedRange<Self>, lable: String = "", nestingLevel: Int = 0) -> Self {
    while true {
      let value = read(lable: lable, nestingLevel: nestingLevel)
      if range.contains(value) {
        return value
      } else {
        print("Error: Value must be in the range '\(range)'")
      }
    }
  }
}

class Transportation {
  enum CargoType {
    case usual
    case oversized
    case fireHazard

    func description() -> String {
      switch self {
      case .usual:
        return "Usual"
      case .oversized:
        return "Oversized"
      case .fireHazard:
        return "Fire hazard"    
      }
    }

    /// Returns a `CargoType` read from standard input.
    static func read(lable: String = "Cargo type", nestingLevel: Int = 0) -> CargoType {
      let indent = String(repeating: " ", count: nestingLevel * 2)
      Swift.print("\(indent)> \(lable)")

      let cargoTypes: [CargoType] = [.usual, .oversized, .fireHazard]
      for (index, cargoType) in cargoTypes.enumerated() {
        Swift.print("\(indent)    \(index + 1). \(cargoType.description())")
      }

      let choise = Int.read(in: 1...cargoTypes.count, nestingLevel: nestingLevel)
      return cargoTypes[choise - 1]
    }
  }

  let cargoType: CargoType
  let cargoMass: Double
  let distance: Int

  init(cargoType: CargoType, cargoMass: Double, distance: Int) {
    self.cargoType = cargoType
    self.cargoMass = cargoMass
    self.distance = distance
  }

  /// Returns a `Transportation` read from standard input.
  static func read(lable: String = "Transportation", nestingLevel: Int = 0) -> Transportation {
    let indent = String(repeating: " ", count: nestingLevel * 2)
    Swift.print("\(indent)> \(lable)")
    return Transportation(
      cargoType: CargoType.read(nestingLevel: nestingLevel + 1),
      cargoMass: Double.read(in: 1...100_000, lable: "Cargo mass (kg)", nestingLevel: nestingLevel + 1),
      distance: Int.read(in: 1...100_000, lable: "Distance (km)", nestingLevel: nestingLevel + 1)
    )
  }

  /// Writes the textual representations of the `Transportation` into the standard output.
  func print(terminator: String = "\n") {
    Swift.print("""
    Transportation:
      Cargo type: \(cargoType.description())
      Cargo mass: \(cargoMass)
      Distance:   \(distance)
    """, terminator: terminator)
  }
}

class Trucks {
  var transportations = [Transportation]()

  func append() {
    transportations.append(Transportation.read(lable: "New transportation"))
  }

  func remove() {
    if transportations.isEmpty {
      Swift.print("Note: Empty")
    } else {
      let index = Int.read(in: 0...(transportations.count - 1), lable: "Index")
      transportations.remove(at: index)
    }
  }
   
  /// Writes the textual representations of the `Trucks` into the standard output.
  func print(terminator: String = "\n") {
    if transportations.isEmpty {
      Swift.print("Note: Empty", terminator: terminator)
    } else {
      let separator = String(repeating: "-", count: 80)
      for transportation in transportations {
        Swift.print(separator)
        transportation.print()
      }
      Swift.print(separator, terminator: terminator)
    }
  }
}

// MARK: Main

print("""
     ______                 __
    /_  __/____ __ __ ____ / /__ ___
     / /  / __// // // __//  '_/(_-<
    /_/  /_/   \\_,_/ \\__//_/\\_\\/___/
""", terminator: "\n\n")

var trucks = Trucks()

menu: while true {
  print("""
  > Menu:
      1. Append
      2. Remove
      3. Print
      4. Exit
  """)
  let choise = Int.read(in: 0...4)
  print() // Empty line.

  switch choise {
  case 0:
    print("  Created with ❤️ by jaroshevskii", terminator: "\n\n")
  case 1:
    trucks.append()
    print() // Empty line.
  case 2:
    trucks.remove()
    print() // Empty line.
  case 3:
    trucks.print(terminator: "\n\n")
  case 4: // Exit.
    print("  (* ^ ω ^) goodbye", terminator: "\n\n")
    break menu
  default:
    fatalError("Invalid choise")
  }
}

File: practice/practical-work7/Main.cpp
//
// Object-oriented programing (college 4th course)
// Practice: Practical work 7
//

#include <iostream>
#include <string>
#include <vector>

int readInt(const std::string &lable = "", int nestingLevel = 0) {
  const std::string indent(nestingLevel * 2, ' ');
  const std::string decor = lable == "" ? "> " : "> " + lable + " = ";
  int value = 0;
  std::cout << indent << decor, std::cin >> value;
  return value;
}

int readInt(int min, int max, const std::string &lable = "",
            int nestingLevel = 0) {
  while (true) {
    if (const auto value = readInt(lable, nestingLevel);
        value >= min && value <= max) {
      return value;
    }
    std::cerr << "Error: Value must be in the range [" << min << "..." << max
              << "].\n";
  }
}

class Transportation {
public:
  enum class CargoType { Usual, Oversized, FireHazard, Unknown };

  static CargoType readCargoType(const std::string &lable = "Cargo type",
                                 int nestingLevel = 1) {
    const std::string indent(nestingLevel * 2, ' ');
    std::cout << indent << "> " << lable << ":\n"
              << indent << "    1. Usual\n"
              << indent << "    2. Oversized\n"
              << indent << "    3. FireHazard\n";
    const int choise = readInt(1, 3, "", nestingLevel);
    switch (choise) {
    case 1:
      return CargoType::Usual;
    case 2:
      return CargoType::Oversized;
    case 3:
      return CargoType::FireHazard;
    default:
      return CargoType::Unknown;
    }
  }

private:
  CargoType cargoType = CargoType::Unknown;
  int cargoMass = 0.0;
  int distance = 0;

public:
  Transportation(const std::string &lable) { read(lable); }

  void read(const std::string &lable) {
    std::cout << "> " << lable << ":\n";
    cargoType = readCargoType();
    cargoMass = readInt(1, 10'000, "Cargo mass (kg)", 1);
    distance = readInt(1, 1'000'000, "Distance", 1);
  }

  std::string cargoTypeToString(const CargoType &cargoType) const {
    switch (cargoType) {
    case CargoType::Usual:
      return "Usual";
    case CargoType::Oversized:
      return "Oversized";
    case CargoType::FireHazard:
      return "Fire hazard";
    case CargoType::Unknown:
      return "Unknown";
    }
  }

  void print(const std::string &lable) const {
    std::cout << lable << ":\n"
              << "  Cargo type: " << cargoTypeToString(cargoType) << '\n'
              << "  Cargo mass: " << cargoMass << '\n'
              << "  Distance:   " << distance << '\n';
  }

  CargoType getCargoType() const { return cargoType; }

  int getCargoMass() const { return cargoMass; }

  int getDistance() const { return distance; }
};

class Trucks {
private:
  std::vector<Transportation> trucksTransportations;

public:
  void add() {
    trucksTransportations.push_back(Transportation("Truck transportation"));
    std::cout << '\n';
  }

  void print(const Transportation::CargoType &cargoType =
                 Transportation::CargoType::Unknown) const {
    const std::string separator(80, '-');
    for (const auto &transportation : trucksTransportations) {
      if (cargoType == Transportation::CargoType::Unknown ||
          cargoType == transportation.getCargoType()) {
        std::cout << separator << '\n';
        transportation.print("Truck transportation");
      }
    }
    std::cout << separator << "\n\n";
  }

  void printSortedByDistanceOfTransportation() const {
    Trucks copy(*this);
    auto &transportations = copy.trucksTransportations;
    for (int i = 0; i < transportations.size(); ++i) {
      for (int j = i + 1; j < transportations.size(); ++j) {
        if (transportations[i].getDistance() >
            transportations[j].getDistance()) {
          std::swap(transportations[i], transportations[j]);
        }
      }
    }
    copy.print();
  }

  void printWithLargerCargoMass(int cargoMass) const {
    const std::string separator(80, '-');
    for (const auto &transportation : trucksTransportations) {
      if (transportation.getCargoMass() > cargoMass) {
        std::cout << separator << '\n';
        transportation.print("Truck transportation");
      }
    }
    std::cout << separator << "\n\n";
  }
};

int main() {
  std::cout << "     ______                 __\n"
            << "    /_  __/____ __ __ ____ / /__ ___\n"
            << "     / /  / __// // // __//  '_/(_-<\n"
            << "    /_/  /_/   \\_,_/ \\__//_/\\_\\/___/\n\n";

  Trucks trucks;

  while (true) {
    std::cout
        << "> Menu:\n"
        << "    1. Add truck\n"
        << "    2. Print trucks\n"
        << "    3. Print trucks with a given type of cargo\n"
        << "    4. Print sorted by distance of transportation\n"
        << "    5. Print trucks which have a mass greater than the given\n"
        << "    6. Exit\n";
    const int choise = readInt(0, 6);
    std::cout << '\n';

    switch (choise) {
    case 0:
      std::cout << "  Created with ❤️ by jaroshevskii\n\n";
      break;
    case 1:
      trucks.add();
      break;
    case 2:
      trucks.print();
      break;
    case 3: {
      const auto cargoType =
          Transportation::readCargoType("Transportation cargo type", 0);
      std::cout << '\n';
      trucks.print(cargoType);
      break;
    }
    case 4:
      trucks.printSortedByDistanceOfTransportation();
      break;
    case 5: {
      const int cargoMass = readInt(1, 10'000, "Cargo mass (kg)");
      std::cout << '\n';
      trucks.printWithLargerCargoMass(cargoMass);
      break;
    }
    case 6: // Exit.
      std::cout << "  (* ^ ω ^) goodbye \n\n";
      return 0;
    }
  }
  return 0;
}
